### PR TITLE
[Sapling] Try to recover corruption of notes cache during send operation

### PR DESCRIPTION
**Pull request by random-zebra**
This is a contingency plan, in case of missing cached nullifier of a spent note in the wallet.
Since, during the spend, the wallet is unlocked, and we have the spending key, we can directly recover the nullifier from the note (provided that we have a valid cached witness) and fix the issue in the background, without requiring any user intervention.

If we don't have the witness, we can still provide a better log and error message in the gui, than a generic, and probably obscure
```
bad-txns-shielded-requirements-not-met
```

### CODE REVIEW

File: src/qt/pivx/send.cpp [hunk 1]
```diff
@@ -449,6 +449,7 @@ void SendWidget::ProcessSend(const QList<SendCoinsRecipient>& recipients, bool h
 
 OperationResult SendWidget::prepareShielded(WalletModelTransaction* currentTransaction, bool fromTransparent)
 {
+    AssertLockHeld(pwalletMain->cs_wallet);
```
Comment (furszy) [inline src/qt/pivx/send.cpp:452]: This isn't being locked on the GUI, nor should been locked.

File: src/qt/walletmodel.cpp [hunk 2]
```diff
@@ -527,6 +527,7 @@ OperationResult WalletModel::PrepareShieldedTransaction(WalletModelTransaction*
                                                         bool fromTransparent,
                                                         const CCoinControl* coinControl)
 {
+    AssertLockHeld(pwalletMain->cs_wallet);
```
Comment (furszy) [inline src/qt/walletmodel.cpp:530]: Same as on `send.cpp`, this isn't being locked on the GUI. GUI side should never lock `cs_wallet` (and that is the future cleanup goal in the places where we do it, like in the txModel).

File: src/sapling/sapling_operation.cpp [hunk 3]
```diff
@@ -65,6 +65,7 @@ TxValues calculateTarget(const std::vector<SendManyRecipient>& recipients, const
 
 OperationResult SaplingOperation::build()
 {
+    AssertLockHeld(pwalletMain->cs_wallet);
```
Comment (furszy) [inline src/sapling/sapling_operation.cpp:68]: This isn't needed. If we have to lock `cs_wallet` due the new changes (like the nullifier set), it should be done inside this method.

File: src/sapling/sapling_operation.cpp [hunk 4]
```diff
@@ -376,6 +377,7 @@ static CacheCheckResult CheckCachedNote(const SaplingNoteEntry& t, const libzcas
 
 OperationResult SaplingOperation::loadUnspentNotes(TxValues& txValues, uint256& ovk)
 {
+    AssertLockHeld(pwalletMain->cs_wallet);
```
Comment (furszy) [inline src/sapling/sapling_operation.cpp:380]: This is the opposite, the internal wallet methods call locks it, this shouldn't be locked for the entire execution.

**Comment by random-zebra** [comment]
Locks updated.

**Review by furszy — DISMISSED** [review]
really nice and straightforward contingency plan while we search for the cause 👍.
  ACK 0dc089693f984c4f87233b991974cbec519cf6f9

**Review by furszy — APPROVED** [review]
Haven't checked it with the corrupted/invalid wallet, pure code review, ACK 9c6bf2d.

**Review by furszy — COMMENTED** [review]
Have tested it with the corrupted wallet and all went well but have to say haven't seen any message of the added logging which is weird.
But well, it's an edge case scenario that we should discover the issue soon anyway (in testnet).
All good for me to go ahead with this one.

**Review by Fuzzbawls — APPROVED** [review]
ACK 9c6bf2da0f4445fa8c6f6e52ddf4e17575b1ad5d

**Pull request merged**